src/dbnode/persist/fs/seek_manager.go

```diff
@@ -912,37 +942,39 @@ func (m *seekerManager) openCloseLoop() {
 			break
 		}
 
-		for _, byTime := range m.seekersByShardIdx {
+		// Record seekers for all shard/block combinations to open and close
+		for shardIdx, byTime := range m.seekersByShardIdx {
 			byTime.RLock()
-			accessed := byTime.accessed
-			byTime.RUnlock()
-			if !accessed {
-				continue
+			byTimeSeekersState := byTime.state
+			// Record seekers for a shard which need to be opened
+			if byTimeSeekersState == byTimeSeekersAccessed {
+				shouldTryOpen = append(shouldTryOpen, byTime)
 			}
-			shouldTryOpen = append(shouldTryOpen, byTime)
-		}
-		m.RUnlock()
-
-		// Try opening any unopened times for accessed seekers
-		for _, byTime := range shouldTryOpen {
-			m.openAnyUnopenSeekersFn(byTime)
-		}
-
-		m.RLock()
-		for shard, byTime := range m.seekersByShardIdx {
-			byTime.RLock()
+			// Record seekers for a shard which need to be closed. A seeker for a block might need to be closed either
+			// because a block falls out of retention window or because a shard leaves a node.
+			// Capture seekers to open and seekers to close under the same byTime.rLock to avoid a race condition
+			// wherein a seeker is opened and closed quickly in succession which results in one iteration of this loop
+			// opening and closing the same seeker resulting in wasted work.
 			for blockStartNano := range byTime.seekers {
 				blockStart := blockStartNano.ToTime()
-				if blockStart.Before(earliestSeekableBlockStart) {
+				if byTime.state == byTimeSeekersClosed || blockStart.Before(earliestSeekableBlockStart) {
 					shouldClose = append(shouldClose, seekerManagerPendingClose{
-						shard:      uint32(shard),
+						shard:      uint32(shardIdx),
 						blockStart: blockStart,
 					})
 				}
 			}
+
 			byTime.RUnlock()
 		}
+		m.RUnlock()
```

Comment:
Ok, so here's the potential lock gap I was mentioning earlier today. 

Previous code does:

```
// handle all opens

m.RLock()
// find seekers outside retention
// filter out any still active seekers
m.RUnlock()

// do the actual close
```

IIUC, your code does:

```
m.RLock()
// find seekers to open 
// find seekers outside retention (and explicitly closed)
m.RUnlock()

// do the actual open

m.RLock()
// filter out any still active seekers
m.RUnlock()
// do the actual close
```

It seems like there's now a locking gap (on m's mutex) between `// find seekers outside retention (and explicitly closed)` and  `// filter out any still active seekers` that didn't exist before. Am I reading things correctly? If so, thoughts? Seems like we at least access `m.seekersByShardIdx` in the "filter" section, so things potentially could have changed there.

Reply:
This gap used to exist even before this PR. 

Thing that has changed now is that, earlier once seekers for a block was closed they were never reopened because the only reason seekers for a block were ever closed was that they fell out of the retention window.

After this PR, seekers can re-open once closed due to outside intervention (CloseShardIndices()) as a result the gap in locking which used to exist became significant after this PR.